### ISSUE TITLE
Stream gif and heic thumbnails

### ISSUE THREAD

**Issue by isaldana** [libvips/libvips#1829]
I am trying to resize and stream .gif files. It works fine when writing to a file but not when streaming.

The following command to write to a file works:

```
$ vipsthumbnail sample_animated.gif -o x1.gif --size '400x>'
```

but the following fails
```
$ cat sample_animated.gif|vips thumbnail_source [descriptor=0] --size down .gif 400 --height 16000 > x2.gif
VipsForeignSave: ".gif" is not a known target format
```

Is there anyway I can save a .gif file to stdout? I am assuming `vipsthumbnail` doesn't support stdout yet.

A similar but maybe unrelated issue occurs with heic files.

The following command works:

```
$ vipsthumbnail sample1.heic -o x.jpg --size '400x>'
```

but the following fails
```
$ cat sample1.heic |vips thumbnail_source [descriptor=0] --size down .png '400x>' --height 16000 > x.jpg
heif: Invalid input: No 'ftyp' box (2.102)
```

**Comment by jcupitt** [comment]
Hello @isaldana,

Sorry, no. GIF write uses imagemagick, and the `magicksave` operator does not support stream output. The best you can do is save to a memory buffer, then copy the buffer to stdout.

**Comment by jcupitt** [comment]
You're right, this should work:

```
$ cat Example1.heic | vips thumbnail_source [descriptor=0] .jpg 128 | cat > x.jpg
heif: Invalid input: No 'ftyp' box (2.102)
```

Though curiously this works:

```
$ cat Example1.heic | vips heifload_source [descriptor=0] x.jpg
```

Some kind of bug in `heifload` I suppose. I'll have a look, thanks for the report!

**Comment by isaldana** [comment]
FWIW, I just noticed that a similar issue occurs with tiff files:

```
$ cat sample.tiff | vips thumbnail_source [descriptor=0] .jpg 128 | cat > x.jpg
tiff2vips: TIFF does not contain page 1
```

But the following works:
```
$ cat sample.tiff | vips tiffload_source [descriptor=0] x.jpg
```

**Comment by kleisauke** [comment]
This looks like a sniff (or rewind) bug. It seems to work properly when the sniffer is skipped in `heifload`. For example:
```diff
diff --git a/libvips/foreign/heifload.c b/libvips/foreign/heifload.c
index 1111111..2222222 100644
--- a/libvips/foreign/heifload.c
+++ b/libvips/foreign/heifload.c
@@ -1171,7 +1171,7 @@ vips_foreign_load_heif_source_build( VipsObject *object )
 		g_object_ref( heif->source );
 	}
 
-	if( VIPS_OBJECT_CLASS( vips_foreign_load_heif_file_parent_class )->
+	if( VIPS_OBJECT_CLASS( vips_foreign_load_heif_source_parent_class )->
 		build( object ) )
 		return( -1 );
 
@@ -1183,8 +1183,7 @@ vips_foreign_load_heif_source_is_a_source( VipsSource *source )
 {
 	const char *p;
 
-	return( (p = (const char *) vips_source_sniff( source, 12 )) &&
-		vips_foreign_load_heif_is_a( p, 12 ) );
+	return( TRUE );
 }
 
 static void
```

Tested with:
```bash
cat test/test-suite/images/Example1.heic | vips thumbnail_source [descriptor=0] x.jpg 128 --vips-info
```

**Comment by kleisauke** [comment]
It seems that the sniffer buffer is not reused, this patch seems to work for me:
```diff
diff --git a/libvips/foreign/heifload.c b/libvips/foreign/heifload.c
index 1111111..2222222 100644
--- a/libvips/foreign/heifload.c
+++ b/libvips/foreign/heifload.c
@@ -1171,7 +1171,7 @@ vips_foreign_load_heif_source_build( VipsObject *object )
 		g_object_ref( heif->source );
 	}
 
-	if( VIPS_OBJECT_CLASS( vips_foreign_load_heif_file_parent_class )->
+	if( VIPS_OBJECT_CLASS( vips_foreign_load_heif_source_parent_class )->
 		build( object ) )
 		return( -1 );
 
diff --git a/libvips/iofuncs/source.c b/libvips/iofuncs/source.c
index 1111111..2222222 100644
--- a/libvips/iofuncs/source.c
+++ b/libvips/iofuncs/source.c
@@ -1262,6 +1262,12 @@ vips_source_sniff( VipsSource *source, size_t length )
 	if( vips_source_test_features( source ) )
 		return( NULL );
 
+	/* Try to reuse the sniffer buffer, if applicable
+	 */
+	if( source->sniff->data &&
+		source->sniff->len >= length )
+		return( source->sniff->data );
+
 	bytes_read = vips_source_sniff_at_most( source, &data, length );
 	if( bytes_read == -1 )
 		return( NULL );
```

I'm not sure if this also resolves the TIFF issue, as I could not reproduce it using this image:
https://github.com/libvips/libvips/blob/master/test/test-suite/images/sample.tif

**Comment by jcupitt** [comment]
I think I've found it! heifload needs to rewind the source before opening it. Other loaders do this, but we seem to have forgotten for this one :(

```
john@yingna ~/GIT/libvips/libvips/foreign (master) $ grep vips_source_rewind *.c
csvload.c:	if( vips_source_rewind( csv->source ) )
csvload.c:	if( vips_source_rewind( csv->source ) )
foreign.c:		(void) vips_source_rewind( source );
gifload.c:	vips_source_rewind( gif->source );
jpeg2vips.c:		if( vips_source_rewind( jpeg->source ) )
matrixload.c:	if( vips_source_rewind( matrix->source ) )
ppmload.c:	if( vips_source_rewind( ppm->source ) )
radiance.c:	if( vips_source_rewind( source ) )
spngload.c:	if( vips_source_rewind( png->source ) ) 
svgload.c:	if( vips_source_rewind( source->source ) )
svgload.c:	if( vips_source_rewind( source->source ) ||
tiff.c:	if( vips_source_rewind( source ) )
vipspng.c:	if( vips_source_rewind( source ) ) 
```

I'll check if any other loaders are also missing a rewind.

foreign.c rewinds the source before calling `_is_a`, but doesn't before calling `_header`, curiously. Perhaps that would be a safer place to put it.

**Comment by jcupitt** [comment]
I found some problems with the way we were faking seek on pipes as well, and also fixed that awful typo you found Kleis.

This improvement will be in 8.10.2.

**Comment by jcupitt** [comment]
I fixed a few other things re. pipes and streams in 8.10 :( Any testing would be great.

git master has a revised vipsthumbnail that can work with pipes and redirects, for example:

```
$ vipsthumbnail stdin -o x.jpg < Example1.heic 
$ cat Example1.heic | vipsthumbnail stdin -o x.jpg
$ cat Example1.heic | vipsthumbnail stdin -o .jpg | cat >x.jpg
```

Hopefully the syntax is a bit more obvious.

**Comment by JackIngleton** [comment]
> git master has a revised vipsthumbnail that can work with pipes and redirects, for example:
> 
> ```
> $ vipsthumbnail stdin -o x.jpg < Example1.heic 
> $ cat Example1.heic | vipsthumbnail stdin -o x.jpg
> $ cat Example1.heic | vipsthumbnail stdin -o .jpg | cat >x.jpg
> ```
> 
> Hopefully the syntax is a bit more obvious.

@jcupitt is this now released?

`cat Example1.heic | jpegsave stdin -o .jpg | cat >x.jpg` does not work for me, whereas `cat Example1.heic | jpegsave stdin /dev/stdout | cat >x.jpg` does.

**Comment by jcupitt** [comment]
Hi @JackIngleton, yes, it should work in 8.11.

I think your second example should be:

```
$ cat Example1.heic | vips copy stdin .jpg | cat >x.jpg
```

`jpegsave` doesn't support `.jpg` as an output image argument. This is because the special `.jpg` output filename conversion is part of the image type.

**Comment by jcupitt** [comment]
Sorry, that wasn't very clear. Try this:

```
 $ vips jpegsave
save image to jpeg file
usage:
   jpegsave in filename [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
   Q            - Q factor, input gint
			default: 75
			min: 1, max: 100
...
```

You can see the second argument to `jpegsave` is just a simple string (the filename to write to), and `jpegsave` will just write to that string no matter what. You need to use `jpegsave_buffer` (for example) if you want to send it somewhere else.

Now try this:

```
$ vips copy
copy an image
usage:
   copy in out [--option-name option-value ...]
where:
   in           - Input image, input VipsImage
   out          - Output image, output VipsImage
optional arguments:
   width        - Image width in pixels, input gint
			default: 0
			min: 0, max: 10000000
...
```

The second argument to `copy` is the output image. The `vips` driver program will use the string you supply for this argument to construct a destination object and pick a saver for you. It's this construction process that handles magic filenames like `.jpg`.

tldr: `jpegsave` is very low-level and does exactly what you tell it to do. `copy` is higher-level and will pick the low-level operations for you.

**Comment by JackIngleton** [comment]
@jcupitt Thank you for the really quick reply!

I actually need to set the quality option on the `jpegsave` command, so I'm assuming `copy` isn't an option here?

`vips jpegsave stdin /dev/stdout -Q 100` is working fine for me anyway, thank you for the clarification.

**Comment by jcupitt** [comment]
You can pass options to the loaders and savers that copy picks for you in square brackets, eg.:

```
$ vips copy x.png y.jpg[Q=90,interlace]
```

**Comment by JackIngleton** [comment]
Great, thank you.
